File: ts/src/checksum/Crc32Calculator.ts
import { Utf8 } from "../encoding/Utf8";
import { Base62Codec } from "../encoding/Base62Codec";
import { Calculator } from "./Calculator";
import { Checksum } from "./Checksum";
import { Crc32 } from "./Crc32";


/**
 * Crc32Calculator derives the tail checksum from the CRC32-IEEE value of a base string encoded as
 * Base62.
 */
export class Crc32Calculator implements Calculator {
    private readonly _encoder: Base62Codec;

    /**
     * Constructs a Crc32Calculator backed by the given Base62 codec.
     */
    public constructor(encoder: Base62Codec) {
        this._encoder = encoder;
    }

    /**
     * Computes the CRC32-IEEE value of the base string interpreted as a UTF-8 byte stream and
     * encodes it as a fixed-width Base62 checksum.
     */
    public calculate(baseString: string): Checksum {
        const value: number = Crc32.checksumIeee(Utf8.encode(baseString));
        const encoded: string = this._encoder.encode(value, Checksum.LENGTH);

        return new Checksum(encoded);
    }
}

File: go/internal/checksum/crc32_calculator.go
package checksum

import (
	"hash/crc32"

	"go.leoweyr.com/keycutter/go/v2/internal/encoding"
)

// Crc32Calculator derives the tail checksum from the CRC32-IEEE value of a base
// string encoded as Base62.
type Crc32Calculator struct {
	encoder *encoding.Base62Codec
}

// NewCrc32Calculator builds a Crc32Calculator backed by the given Base62 codec.
func NewCrc32Calculator(encoder *encoding.Base62Codec) *Crc32Calculator {
	return &Crc32Calculator{encoder: encoder}
}

// Calculate computes the CRC32-IEEE value of the base string interpreted as a UTF-8
// byte stream and encodes it as a fixed-width Base62 checksum.
func (crc32Calculator *Crc32Calculator) Calculate(baseString string) *Checksum {
	var value uint32 = crc32.ChecksumIEEE([]byte(baseString))
	var encoded string = crc32Calculator.encoder.Encode(value, Length)

	return NewChecksum(encoded)
}
